api/v1alpha1/serviceresolver_types_test.go

```diff
@@ -0,0 +1,509 @@
+package v1alpha1
+
+import (
+	"testing"
+	"time"
+
+	capi "github.com/hashicorp/consul/api"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Test MatchesConsul for cases that should return true.
+func TestServiceResolver_MatchesConsulTrue(t *testing.T) {
+	cases := map[string]struct {
+		Ours   ServiceResolver
+		Theirs *capi.ServiceResolverConfigEntry
+	}{
+		"empty fields": {
+			Ours: ServiceResolver{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "name",
+				},
+				Spec: ServiceResolverSpec{},
+			},
+			Theirs: &capi.ServiceResolverConfigEntry{
+				Name: "name",
+				Kind: capi.ServiceResolver,
+			},
+		},
+		"all fields set": {
+			Ours: ServiceResolver{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "name",
+				},
+				Spec: ServiceResolverSpec{
+					DefaultSubset: "default_subset",
+					Subsets: map[string]ServiceResolverSubset{
+						"subset1": {
+							Filter:      "filter1",
+							OnlyPassing: true,
+						},
+						"subset2": {
+							Filter:      "filter2",
+							OnlyPassing: false,
+						},
+					},
+					Redirect: &ServiceResolverRedirect{
+						Service:       "redirect",
+						ServiceSubset: "redirect_subset",
+						Namespace:     "redirect_namespace",
+						Datacenter:    "redirect_datacenter",
+					},
+					Failover: map[string]ServiceResolverFailover{
+						"failover1": {
+							Service:       "failover1",
+							ServiceSubset: "failover_subset1",
+							Namespace:     "failover_namespace1",
+							Datacenters:   []string{"failover1_dc1", "failover1_dc2"},
+						},
+						"failover2": {
+							Service:       "failover2",
+							ServiceSubset: "failover_subset2",
+							Namespace:     "failover_namespace2",
+							Datacenters:   []string{"failover2_dc1", "failover2_dc2"},
+						},
+					},
+					ConnectTimeout: 1 * time.Second,
+				},
+			},
+			Theirs: &capi.ServiceResolverConfigEntry{
+				Name:          "name",
+				Kind:          capi.ServiceResolver,
+				DefaultSubset: "default_subset",
+				Subsets: map[string]capi.ServiceResolverSubset{
+					"subset1": {
+						Filter:      "filter1",
+						OnlyPassing: true,
+					},
+					"subset2": {
+						Filter:      "filter2",
+						OnlyPassing: false,
+					},
+				},
+				Redirect: &capi.ServiceResolverRedirect{
+					Service:       "redirect",
+					ServiceSubset: "redirect_subset",
+					Namespace:     "redirect_namespace",
+					Datacenter:    "redirect_datacenter",
+				},
+				Failover: map[string]capi.ServiceResolverFailover{
+					"failover1": {
+						Service:       "failover1",
+						ServiceSubset: "failover_subset1",
+						Namespace:     "failover_namespace1",
+						Datacenters:   []string{"failover1_dc1", "failover1_dc2"},
+					},
+					"failover2": {
+						Service:       "failover2",
+						ServiceSubset: "failover_subset2",
+						Namespace:     "failover_namespace2",
+						Datacenters:   []string{"failover2_dc1", "failover2_dc2"},
```

Comment:
It would be considered different because we're using `reflect.DeepEqual`. We could look at using a different comparator but in practice they would only get reordered if someone were to edit them so I think it might be okay.

Ashwin and I just talked about not implementing `MatchesConsul` and instead using `ToConsul()` and then reflect.DeepEqual so we don't have to write as much code.

Reply:
That's fair since this param is an array, and so the order should matter.